File: sdk/csharp/SymbolSdk/Nem/SharedKey.cs
using Org.BouncyCastle.Crypto.Digests;

namespace SymbolSdk.Nem;

public abstract class SharedKey
{
    private static void CryptoHash(byte[] outArray, byte[] m) {
        var digest = new KeccakDigest(512);
        digest.BlockUpdate(m, 0, m.Length);
        digest.DoFinal(outArray, 0);
    }

    public static SharedKey256 DeriveSharedKey(byte[] privateKey, byte[] otherPublicKey) {
        var deriveSharedKeyImpl = SymbolSdk.SharedKey.DeriveSharedKeyFactory("nem-nis1", CryptoHash);
        return deriveSharedKeyImpl(privateKey.Reverse().ToArray(), otherPublicKey);
    }
}

File: sdk/dart/lib/nem/SharedKey.dart
import 'dart:typed_data';
import 'package:symbol_sdk/index.dart';

import '../CryptoTypes.dart';
import 'package:pointycastle/export.dart' as pointy_castle;

void cryptoHash(Uint8List out, Uint8List m) {
  var digest = pointy_castle.KeccakDigest(512);
  var r = digest.process(m);

  for (var i = 0; i < out.length; ++i) {
    out[i] = r[i];
  }
}

final deriveSharedKeyImpl = deriveSharedKeyFactory('nem-nis1', cryptoHash);

SharedKey256 deriveSharedKey(dynamic keyPair, PublicKey otherPublicKey) {
  return deriveSharedKeyImpl(Uint8List.fromList(keyPair.privateKey.bytes.reversed.toList()), otherPublicKey);
}
